drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2149 @@
+
+	(void)pm_device_runtime_put(dev);
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+}
+
+#ifdef CONFIG_I3C_USE_IBI
+
+int i3c_stm32_ibi_enable(const struct device *dev, struct i3c_device_desc *target)
+{
+	int ret = 0;
+	uint8_t idx;
+	I3C_TypeDef *i3c;
+	struct i3c_ccc_events i3c_events;
+	struct i3c_stm32_data *data = dev->data;
+	const struct i3c_stm32_config *config = dev->config;
+
+	i3c = config->i3c;
+	if (!i3c_device_is_ibi_capable(target)) {
+		return -EINVAL;
+	}
+
+	if (data->ibi.num_addr >= ARRAY_SIZE(data->ibi.addr)) {
+		/* No more free entries in the IBI table */
+		LOG_ERR("%s: no more free space in the IBI table", __func__);
+		return -ENOMEM;
+	}
+
+	for (idx = 0; idx < ARRAY_SIZE(data->ibi.addr); idx++) {
+		if (data->ibi.addr[idx] == target->dynamic_addr) {
+			LOG_ERR("%s: selected target is already in the list", __func__);
+			return -EINVAL;
+		}
+	}
+
+	if (data->ibi.num_addr > 0) {
+		for (idx = 0; idx < ARRAY_SIZE(data->ibi.addr); idx++) {
+			if (data->ibi.addr[idx] == 0U) {
+				break;
+			}
+		}
+
+		if (idx >= ARRAY_SIZE(data->ibi.addr)) {
+			LOG_ERR("Cannot support more IBIs");
+			return -ENOTSUP;
+		}
+
+	} else {
+		idx = 0;
+	}
+
+	data->ibi.addr[idx] = target->dynamic_addr;
+	data->ibi.num_addr += 1U;
+
+	if (data->ibi.num_addr == 1U) {
+		(void)pm_device_runtime_get(dev);
+	}
+
+	/* Tell target to enable IBI */
+	i3c_events.events = I3C_CCC_EVT_INTR;
+	ret = i3c_ccc_do_events_set(target, true, &i3c_events);
+	if (ret != 0) {
+		LOG_ERR("Error sending IBI ENEC for 0x%02x (%d)", target->dynamic_addr, ret);
+	}
+
+	/* Set I3C bus devices configuration */
+	LL_I3C_ConfigDeviceCapabilities(i3c, (idx + 1), target->dynamic_addr,
+					LL_I3C_IBI_CAPABILITY, LL_I3C_IBI_DATA_ENABLE,
+					LL_I3C_CR_NO_CAPABILITY);
+
+	return ret;
+}
+
+int i3c_stm32_ibi_disable(const struct device *dev, struct i3c_device_desc *target)
+{
+	int ret = 0;
+	uint8_t idx;
+	I3C_TypeDef *i3c;
+	struct i3c_ccc_events i3c_events;
+	struct i3c_stm32_data *data = dev->data;
+	const struct i3c_stm32_config *config = dev->config;
+
+	i3c = config->i3c;
+	if (!i3c_device_is_ibi_capable(target)) {
+		return -EINVAL;
+	}
+
+	for (idx = 0; idx < ARRAY_SIZE(data->ibi.addr); idx++) {
+		if (target->dynamic_addr == data->ibi.addr[idx]) {
+			break;
+		}
+	}
+
+	if (idx == ARRAY_SIZE(data->ibi.addr)) {
+		LOG_ERR("%s: target is not in list of registered addresses", __func__);
+		return -ENODEV;
+	}
+
+	data->ibi.addr[idx] = 0U;
+	data->ibi.num_addr -= 1U;
+
+	if (data->ibi.num_addr == 0U) {
+		(void)pm_device_runtime_put(dev);
+	}
+
+	/* Tell target to disable IBI */
+	i3c_events.events = I3C_CCC_EVT_INTR;
+	ret = i3c_ccc_do_events_set(target, false, &i3c_events);
+	if (ret != 0) {
+		LOG_ERR("Error sending IBI DISEC for 0x%02x (%d)", target->dynamic_addr, ret);
+	}
+
+	if (((target != NULL) && i3c_device_is_ibi_capable(target)) ||
+	    (((target->bcr & I3C_BCR_IBI_REQUEST_CAPABLE) == I3C_BCR_IBI_REQUEST_CAPABLE))) {
```

Comment:
This is checking the same thing twice
you only just need to call `if ((target != NULL) && i3c_device_is_ibi_capable(target))`